src/main/java/org/opentripplanner/transit/model/timetable/ScheduledTripTimes.java

```diff
@@ -0,0 +1,346 @@
+package org.opentripplanner.transit.model.timetable;
+
+import static org.opentripplanner.transit.model.timetable.ValidationError.ErrorCode.NEGATIVE_DWELL_TIME;
+import static org.opentripplanner.transit.model.timetable.ValidationError.ErrorCode.NEGATIVE_HOP_TIME;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.BitSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+import org.opentripplanner.framework.i18n.I18NString;
+import org.opentripplanner.framework.lang.IntUtils;
+import org.opentripplanner.model.BookingInfo;
+import org.opentripplanner.transit.model.basic.Accessibility;
+import org.opentripplanner.transit.model.framework.Deduplicator;
+
+final class ScheduledTripTimes implements Serializable, Comparable<ScheduledTripTimes> {
+
+  /**
+   * Implementation notes: This allows re-using the same scheduled arrival and departure time
+   * arrays for many ScheduledTripTimes. It is also used in materializing frequency-based
+   * ScheduledTripTimes.
+   */
+  private final int timeShift;
+  private final int serviceCode;
+  private final int[] arrivalTimes;
+  private final int[] departureTimes;
+  private final BitSet timepoints;
+  private final Trip trip;
+  private final List<BookingInfo> dropOffBookingInfos;
+  private final List<BookingInfo> pickupBookingInfos;
+
+  @Nullable
+  private final I18NString[] headsigns;
+
+  /**
+   * Implementation notes: This is 2D array since there can be more than one via name/stop per each
+   * record in stop sequence). Outer array may be null if there are no vias in stop sequence. Inner
+   * array may be null if there are no vias for particular stop. This is done in order to save
+   * space.
+   */
+  @Nullable
+  private final String[][] headsignVias;
+
+  private final int[] originalGtfsStopSequence;
+
+  ScheduledTripTimes(ScheduledTripTimesBuilder builder) {
+    this.timeShift = builder.timeShift;
+    this.serviceCode = builder.serviceCode;
+    this.arrivalTimes = builder.arrivalTimes;
+    this.departureTimes = builder.departureTimes;
+    this.timepoints = builder.timepoints;
+    this.trip = builder.trip;
+    this.pickupBookingInfos = builder.pickupBookingInfos;
+    this.dropOffBookingInfos = builder.dropOffBookingInfos;
+    this.headsigns = builder.headsigns;
+    this.headsignVias = builder.headsignVias;
+    this.originalGtfsStopSequence = builder.originalGtfsStopSequence;
+  }
```

Comment:
I think it would be a good idea to add some invariant checks here. Or are there none?

Reply:
Yes, that is the plan - and by splitting TR and schedules times the checks can differ.